core/src/entry.rs

```diff
@@ -218,27 +219,142 @@ pub trait EntrySlice {
     fn to_single_entry_shared_blobs(&self) -> Vec<SharedBlob>;
 }
 
+#[cfg(not(feature = "cuda"))]
+fn verify_entries_scalar(entries: &[Entry], start_hash: &Hash) -> bool {
+    let genesis = [Entry {
+        num_hashes: 0,
+        hash: *start_hash,
+        transactions: vec![],
+    }];
+    let entry_pairs = genesis.par_iter().chain(entries).zip(entries);
+    entry_pairs.all(|(x0, x1)| {
+        let r = x1.verify(&x0.hash);
+        if !r {
+            warn!(
+                "entry invalid!: x0: {:?}, x1: {:?} num txs: {}",
+                x0.hash,
+                x1.hash,
+                x1.transactions.len()
+            );
+        }
+        r
+    })
+}
+
+#[cfg(not(feature = "cuda"))]
+fn verify_entries_simd(
+    entries: &[Entry],
+    start_hash: &Hash,
+    poh_verify_simd: unsafe extern "C" fn(*mut u8, *const u64) -> (),
+    num_per_simd: usize,
+) -> bool {
+    if entries.len() == 0 {
+        return true;
+    }
+
+    let genesis = [Entry {
+        num_hashes: 0,
+        hash: *start_hash,
+        transactions: vec![],
+    }];
+
+    let extra = num_per_simd - (entries.len() % num_per_simd);
+
+    let mut hashes: Vec<Hash> = genesis
+        .par_iter()
+        .chain(entries)
+        .map(|entry| entry.hash)
+        .take(entries.len())
```

Comment:
?

Reply:
What exactly is the question here?

Reply:
What's happening here? There's no code comment and looks quite bizarre. Like you're dropping the last entry, but only because genesis happens to have one entry?  Any what's this doing after the `map()`?  Where are the unit tests for this function?